pkg/storage/etcd.go

```diff
@@ -36,37 +37,77 @@ func NewEtcdStorage(endpoints []string) (*EtcdStorage, error) {
 }
 
 // Store stores the key
-func (etcds *EtcdStorage) Store(key string, value Data) error {
+func (etcds *EtcdStorage) StoreCheckin(key string, value time.Time) error {
 	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
-	encoded, _ := json.Marshal(value)
-	_, err := etcds.kv.Put(ctx, fmt.Sprintf("env=%s", key), string(encoded))
+	encoded, err := json.Marshal(value)
+	if err != nil {
+		cancel()
+		return errors.Wrap(err, "failed to store key")
+	}
+	key = strings.TrimSuffix(strings.TrimPrefix(key, "env/"), "/last") // sanitize the strings
+	_, err = etcds.kv.Put(ctx, fmt.Sprintf("env/%s/last", key), string(encoded))
+	cancel()
+	return errors.Wrap(err, "failed to store key")
+}
+
+func (etcds *EtcdStorage) StoreIncident(key string, date string, value []time.Time) error {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	encoded, err := json.Marshal(value)
+	if err != nil {
+		cancel()
+		return errors.Wrap(err, "failed to store key")
+	}
+	key = strings.TrimSuffix(strings.TrimPrefix(key, "env/"), "/last") // sanitize the strings
+	_, err = etcds.kv.Put(ctx, fmt.Sprintf("env/%s/incidents/%s", key, date), string(encoded))
 	cancel()
 	return errors.Wrap(err, "failed to store key")
 }
 
 // All returns a map of all keys and values
-func (etcds *EtcdStorage) All() (map[string]Data, error) {
+func (etcds *EtcdStorage) AllCheckins() (map[string]time.Time, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+	v, err := etcds.kv.Get(ctx, "env", clientv3.WithPrefix())
+	cancel()
+	if err != nil {
+		return map[string]time.Time{}, errors.Wrap(err, "failed to retrieve all checkins")
+	}
+	all := map[string]time.Time{}
+	for _, item := range v.Kvs {
+		if strings.Contains(string(item.Key), "last") {
+			var data time.Time
+			if err := json.Unmarshal(item.Value, &data); err != nil {
+				return map[string]time.Time{}, errors.Wrap(err, "failed to retrieve all checkins")
+			}
+			all[string(item.Key)] = data
+		}
+	}
+	return all, nil
+}
+
+func (etcds *EtcdStorage) AllIncidents() (map[string][]string, error) {
 	ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
 	v, err := etcds.kv.Get(ctx, "env", clientv3.WithPrefix())
 	cancel()
 	if err != nil {
-		return map[string]Data{}, errors.Wrap(err, "failed to retrieve all stored values")
+		return map[string][]string{}, errors.Wrap(err, "failed to retrieve all stored environments")
 	}
-	all := map[string]Data{}
+	all := map[string][]string{}
 	for _, item := range v.Kvs {
-		var data Data
-		if err := json.Unmarshal(item.Value, &data); err != nil {
-			return map[string]Data{}, nil
+		if strings.Contains(string(item.Key), "incidents") {
+			var data []string
+			if err := json.Unmarshal(item.Value, &data); err != nil {
+				return map[string][]string{}, errors.Wrap(err, "failed to retrieve all stored incidents")
+			}
+			all[fmt.Sprintf("%s", string(item.Key))] = data
```

Comment:
just use `string(item.key)` without the `fmt.Sprintf` as in the previous method.